### ISSUE TITLE
[PSA] "reject" state semantics

### ISSUE THREAD

**Issue by vgurevich** [p4lang/p4-spec#339]
P4_16 spec leaves the semantics of the "reject" state up to the architecture, so we need to define it.

Options:
1. Packets that reach "reject" state are dropped (with an optional counter or counters per error)
2. Packets that reach "reject" state are not dropped, but instead proceed to the ingress control that can choose to drop a packet, ingrement a counter, etc., based on the parser error that should be an input).

**Comment by vgurevich** [comment]
As per brief discussion in the committee...

Option 1 is really simple and protects "unsuspecting" programs from malformed packets. However, given that the packets can reach "reject" state implicitly, it leaves no choices to the advanced users. 

Option 2 is more complicated and the programmer must write the code to deal with the malformed packets. However, the flexibility is enormously better.  For programmers who want to write safe code, while not having to deal with the complexity, PSA can always provide a global action, such as
```
action psa_packet_rejected_drop() {
    if (input_meta.parser_error != error.NoError) {
        pre.drop()
        exit;
    }
}
```

**Comment by jafingerhut** [comment]
An issue from p4lang/p4c that has some related discussion. 
 Not sure if it helps inform the answer to this issue or not: https://github.com/p4lang/p4c/issues/675

**Comment by DGeorgeNetro** [comment]
I quite like Option 2, but if this is the standard it means packets deemed bad by the parser will make it into ingress processing. There might well be some overhead in transitioning between parser and ingress which would could have been avoided for the case of the bad packet.

That said, I don't see any spec-conforming solution other than 2.

**Comment by vgurevich** [comment]
@hanw , @cc10512 ,

I know I am on the hook to write this up, but I think this is actually a very small piece. The most important thing is to decide how errors will be exposed to the ingress control -- without that I can't explain how to process packets that had  parsing/checksum errors.

**Comment by cc10512** [comment]
Please add what you need to the `psa_ingress_input_metadata_t` struct.

**Comment by vgurevich** [comment]
@hanw and other interested parties...

We need to nail down the error handling otherwise ingress (or egress) control will not be able to do much. This consists of the following:
     1. We need to decide how to pass errors into the control. Currently, it is an unspecified bit string (ParserStatus_t) that has a totally unspecified relationship to the `error` type, especially if someone decides to add their own errors. I propose we change the type of `parser_status` to `error` (and potentially rename it to `parser_error`).  This will allow people to define their own errors and pass them consistently. The section on the limits (ho many custom errors can be defined) will need to be updated.
     2. One common error type is IPv4 header checksum error. If we adopt the scheme above, we will have almost no choice but have the check done in the parser state (rather than a separate control). I think this is actually a good thing to do. (Recomputing the checksum can still be done in a separate control or it can be moved into deparser). But there might be some objections: for example, very often checksum verification is somewhat disconnected from parsing. The parser code might show that if there is an IP header checksum error, than no more headers will be extracted, whereas on the real HW, more headers can be extracted before or in parallel with checksum calculation. Personally, I do not think that is a big deal as long as we say that if a packet has any parsing error, the best thing to do is **not** to make any assumptions about it and its headers.
    3. Another implication (if we decide to use (1)) is that only one error can be passed. I do not think this is a serious limitation (ultimately, I would not recommend relying on anything in the packet if a parser error occurred), but others might think otherwise.
    4. I would also recommend either removing parser_error_location or making it optional and platform-dependent, i.e. it can contain a platform-specific indicator of where a transition into the reject state happened (e.g. a byte offset inside the packet, a byte offset from the last extracted header, etc.), but I would not mandate anything. The default value is 0, and nothing should be assumed unless `parser_error` is set to a real error.

Once we finalize these 4 points, the semantics of the reject state becomes clear: the packet is parser into the subsequent control (Ingress or Egress) and `parser_error` is guaranteed to be anything but NoError. The easiest thing to do to deal with such packets would be:
```
apply {
    if (istd.parser_error != error.noError) {
        /* drop, copy to CPU, etc. */
        exit;
    }
    /* Perform normal processing */
}
```

Last, but not least, the scheme, described above should be applicable to both ingressand egress parser.

If we agree on these, we can start reworking the spec.

**Comment by hanw** [comment]
@vgurevich I made a PR on parser error definition with your suggestion. #391

**Comment by jafingerhut** [comment]
A PR with example code for detecting an IPv4 header checksum error in the ingress parser, and using that error value in the ingress control block to drop the packet, is here: https://github.com/p4lang/p4-spec/pull/400

**Comment by jafingerhut** [comment]
PR #400 has been merged in.

parser_error_location has been removed from the latest PSA draft spec as of this writing.  Does anyone want to add it back in, with the caveats mentioned in @vgurevich's earlier comment?  If no one wants to add it back in, then it is probably best to delete the typedef of type ParserErrorLocation_t from psa.p4 that is currently there (but not used anywhere).

There is nothing in the PSA about item number 2 in @vgurevich's earlier comment, in particular the part that I quote below:

"But there might be some objections: for example, very often checksum verification is somewhat disconnected from parsing. The parser code might show that if there is an IP header checksum error, than no more headers will be extracted, whereas on the real HW, more headers can be extracted before or in parallel with checksum calculation. Personally, I do not think that is a big deal as long as we say that if a packet has any parsing error, the best thing to do is not to make any assumptions about it and its headers."

Vladimir, this sounds like an example of making the behavior less defined than the P4_16 v1.0.0 spec currently allows.  Do you think there should be an issue opened on the P4_16 language specification to explicitly allow this kind of undefined reordering of parser operations for packets with parser errors?

**Comment by vgurevich** [comment]
@jafingerhut ,

First of all, thanks a lot for all the work you've done on this!

I think my comment should be interpreted as "thinking out loud". As I said, I do not think this is a big deal, meaning that I think PSA and language semantics should stay as they are. If someone has HW that will be doing some of the things I mentioned, the vendor can easily educate their users, but I think that the vast majority of programs (if not all) will continue to operate on such a device just fine.

If we want to be on the safest side, we can say something like this: 

Option 1: PSA-compliant programs should not assume that any of the headers are valid or invalid if parser_error is set to anything other than NoError. Note: this does not preclude *reading* the header fields: the results will be indeterminate, but it is safe to do. But writing the fields or doing the match will be a bad idea.

Option 2: PSA-compliant programs should not assume that any of the headers are valid or invalid for the standard errors, defined in core.p4. For user-defined errors that are generated by verify statement, PSA-compliant programs can assume that the headers extracted before the verify() statement was hit are valid and should make any assumptions for the other headers.

Option 3: PSA-compliant programs are allowed to assume that all headers that were successfully extracted before the transition to the reject state happened can be assumed valid and it should make no assumptions about other headers. Note: By design, PSA does not provide any mechanisms that would allow pinpointing the exact location where the transition to the reject statement took place. While the logical location can be determined by program analysis, the exact location can vary, based on the specifics of the implementation.

Any thoughts?

**Comment by mbudiu-vmw** [comment]
The user can define an enum to manually keep track of the last state executed.

**Comment by vgurevich** [comment]
@mbudiu-bfn 

Good point!

**Comment by jafingerhut** [comment]
@vgurevich Suppose someone wants to write a P4 program that sends packets that experienced parser errors to the control CPU, with a P4-program-defined header in front of it that contains extra information about the packet.  Do you expect that the programmer can send such a packet through the Ingress control block, doing nothing there except making a "to_control_CPU" header valid and initializing its fields, setting the ostd.egress_port to the CPU_PORT, and then running the packet through the ingress deparser?

Asked another way, when you say that the programmer should not assume anything about which headers are valid or invalid, can they at least assume that the parser leaves things in a state where passing the packet through an appropriate deparser (i.e. one that emit's all valid headers in the order they were parsed) will leave the packet as it was before parsing began?

If they can't even assume _anything_ about the header valid bits, then maybe such a deparser won't even reconstruct the original packet.  Is that what you are proposing?

**Comment by vgurevich** [comment]
@jafingerhut ,

I think the safest way to deliver a bad packet to the CPU is:
   1. Do ingress2egress cloning. This creates an exact copy of the whole packet and avoids ingress parsing
   2. In the ingress deparser emit some data to indicate that this clone will go to the CPU
   3. Do not parse that packet in the egress parser to avoid hitting an exception. That way the whole packet, no matter how wrong will be just the payload
   4. In the egress control, add the CPU header and voila!

Another way to do it would be to resubmit (with some additional data indicating that the packet should not be parsed on the second pass) and then add CPU header in the ingress control. 

Doing it with the parsed packet in the ingress control invites all the problems you just articulated. I am sure it is still viable in many cases, especially if you adopt option 3, but I am sure there will be some corner cases...

**Comment by jafingerhut** [comment]
@vgurevich - Some comments/questions on your proposal.  Also mentioning @hanw to call his attention to these details, as he and I are trying to work through some of the details of exactly how cloning works in PSA.

1. I guess technically, cloning of the original (before-parsing) packet doesn't avoid ingress parsing (i.e. ingress parsing still occurs on the packet).  I think you mean that the packet sent to the control CPU won't be modified in any way by whatever ingress parsing does?

2. The current PSA draft has pseudocode specifying that a clone is generated after ingress, if the Ingress control block sets the 'clone' field to true in the struct with type 'psa_ingress_output_metadata_t' that is an 'out' parameter of the Ingress control block.  Also in the current PSA draft, the deparsers (both ingress and egress) can only take the parsed version of the headers as input, and produce a packet_out as output.  They cannot view or modify metadata.  Do you think that should be changed?

3. How does the egress deparser "know" to avoid parsing this packet?  Because it is a CLONE type, not NORMAL, and it has egress_port=CPU_PORT?  What if we want to add a header to the to-CPU packet that contains the original ingress_port where the packet arrived?

4. The egress does not currently get that value in the metadata, only egress_port and a few other egress-related fields.  Should every egress packet have the original ingress_port as a metadata field?  Or should the ingress somehow add a header before the original packet received that contains whatever the P4 program wants to include there?

**Comment by jafingerhut** [comment]
I have added a few sentences in this PR based upon @mbudui-vmw's comment of how a P4 programmer can explicitly record information about where a parser error occurred, if they wish: https://github.com/p4lang/p4-spec/issues/339#issuecomment-325724452

**Comment by vgurevich** [comment]
@jafingerhut ,

 1.  I would think that when you clone ingress packet to egress, you take the packet before ingress parsing and shove it into PRE (i.e. after ingress deparsing), because you really want the exact copy of that packet without any possibility of it being changed by ingress processing. I am sure there can be other options, but if ingress mirroring and CPU trapping are two main applications of ingress2egress cloning, that's how I'd like it to be done.

2. I would say that "the clone is generated after ingress *processing of the **original** packet* (has been finished)". So, the metadata is correct. In terms of the deparsing, packet_out refers just to the buffer associated with the original packet. The idea for cloning is that there will be another extern (e.g. clone_out) that will represent the byte stream into which we can emit some data that will be prepended to the cloned packet, presumably to that original bytestream. 
 
3. Correct. 

4. All this can be passed as the additional metadata by emitting into the clone_out extern. Or it can be bridged metadata -- something we haven't yet discussed, but should.

**Comment by jafingerhut** [comment]
@vgurevich @hanw 

2. From where are you expecting that the clone_out extern will have its methods called?
(a) restricted to be only from the Ingress control block?   (or Egress control block for packets cloned on ingress)
(b) restricted to be only from the ingress deparser?  (or egress deparser control block, for packets cloned on egress)
(c) either (a) or (b), as the P4 programmer chooses?
(d) something else?

If (b), then are you expecting that an ingress deparser might look something like this?
```
// The following code is written as if the struct `ostd` that is an
// `out` direction parameter of type `psa_ingress_output_metadata_t`
// from the Ingress control block, is an input to the deparser.  It is
// not that way in the latest PSA spec draft as of this writing.

// The ingress deparser control block is run for all ingress packets,
// after the Ingress control block is complete, but only a subset of
// the code here has an effect, depending upon the type of packet
// being deparsed.

// If a packet is cloned, the ingress deparser is effectively run once
// for the cloned copy, and once for any other type of packet created
// (e.g. multicast, unicast, or resubmitted).

control DeparserImpl(packet_out packet,
                     in headers hdr,
                     in psa_ingress_output_metadata_t ostd)
{
    apply {
        if (ostd.clone) {
            clone.emit(hdr.to_cpu_header);
        }

        if (ostd.clone && ostd.clone_type == CloneType_t.ORIGINAL) {
            // The packet consists of any data emitted by clone.emit()
            // calls above, followed by the original packet bytes as
            // received by the parser.
        } else {
            packet.emit(hdr.ethernet);
            packet.emit(hdr.ipv4);
        }
    }
}
```

**Comment by jafingerhut** [comment]
@vgurevich Regarding bridged metadata, let me know when/where you think would be appropriate to discuss it.  I don't recall hearing of it before, and as far as I know it isn't mentioned in the PSA spec draft yet.

**Comment by vgurevich** [comment]
@jafingerhut ,

You are totally correct, meaning is that the vision is to emit clone_out in the deparser (similar idea is for resubmit_out if we get there, etc.). Another option is to have separate controls, but this looks a little too heavy to me at the moment. 

In terms of the deparser code, the second if() is not needed. In other words it looks like this:
```
control DeparserImpl(packet_out packet,
                     clone_out clone_info,
                     in headers hdr,
                     in psa_ingress_output_metadata_t ostd)
{
    apply {
        if (ostd.clone) {
            clone_info.emit(hdr.to_cpu_header);
            // This will be prepended to the ORIGINAL packet, of clone_type is ORIGINAL or 
            // to the deparsed packet below, but in any case there is no need to do anything else
        }

        // Take care of the original packet as usual, regardless of whether or not it was cloned.
         packet.emit(hdr.ethernet);
         packet.emit(hdr.ipv4);
    }
}
```

**Comment by vgurevich** [comment]
@jafingerhut ,

I'll open another issue for Bridged metadata. It didn't appear, because it is very subtle and got carried over through P4_14 thinking.

**Comment by jafingerhut** [comment]
@vgurevich In your deparser code, I am missing something pretty basic.

If the packet is not cloned, then here is what I thought should be the behavior: the packet.emit() calls use the value of hdr.ethernet.isValid() and hdr.ipv4.isValid() to determine whether to copy data from the fields of hdr.ethernet and hdr.ipv4 headers, _perhaps modified by the Ingress control block code_.  They might emit the headers of the original packet that was given to the ingress parser, or they might emit different headers.

If you want a cloned ingress2egress packet to contain the original packet given to the ingress parser, then how is it possibly correct to execute those packet.emit() calls?

**Comment by hanw** [comment]
The path that the cloned packet travels is different from the path that the normal packet travels. In the case that a cloned ingress2egress packet only contains the original packet given to the ingress parser, the cloned packet is copied directly from the input packet buffer to the PRE (bypassing the ingress pipeline), any field emitted with the `clone.emit()` should be treated as extra metadata to be prepended to the cloned packet (which include the `hdr.ethernet` or `hdr.ipv4` if user chose to emit them in the clone extern).

**Comment by jafingerhut** [comment]
@vgurevich Perhaps.  When you say "`packet_out` represents the bytestream for the original packet only", do you mean that this is the case for cloned packets with clone_type==CloneType_t.ORIGINAL?

Or do you mean that this is also true for non-cloned packets?

If it is also true for non-cloned packets, then what is the point of having in ingress deparser, if every packet_out is always equal to the packet given to the ingress parser as input?

**Comment by jafingerhut** [comment]
OK, I think you should disregard my previous comment.  By "the original packet" in your comment, you mean "the non-cloned packet that comes out of ingress processing, sent to the packet buffer".  You don't mean "the contents of the packet as given to the ingress parser, before ingress processing".  If I have that correct, I was confused by your use of the term "original", and the clone_type==CloneType_t.ORIGINAL condition.

**Comment by jafingerhut** [comment]
@hanw Thanks for that clarification.  I think it would be very good to have a version of what you wrote in the PSA specification very explicitly and clearly.  I think it is a very easy source of confusion that the ingress deparser has this mix of "some statements that only apply to cloned packets" and "other statements that only apply to non-cloned packets".

**Comment by vgurevich** [comment]
@jafingerhut Sorry, I should've used another term. Perhaps a "clone source" or "clone root" or "clonee" (i.e. someone being cloned) would do?

**Comment by jafingerhut** [comment]
@vgurevich @cc10512 @hanw @DGeorgeNetro @mbudiu-vmw 

I have scanned through this discussion, and it has spawned several other separate issues being created, which are being worked on in their own right (e.g. bridged metadata).

As far as I can see, there is nothing left to do for this issue itself, and I recommend we close it perhaps after creating any other issues that should be created from any 'loose ends' not yet addressed, but mentioned above.

**Comment by cc10512** [comment]
Agreed.
